Projectile effects

ElArGee opened the pull request:
**Rewrite of projectile effects:**
- Projectiles fired from a gun now fly in a true straight line (applies to all ranged weapons including beams and bullets)
- Laser effects revamped to be more awesome
- Added muzzle flash for ballistic guns (without suppressors)
- Beam effects are now actual objects instead of overlays. The objects are null spaced and removed by GC instead of being force deleted.

**Note:** This does not change bullets to be hitscan. I decided to drop that since it'll be easy to change later now anyway.

**Pictures:**
![new_laser](https://cloud.githubusercontent.com/assets/6357112/6318087/1f60da5e-bae2-11e4-957e-fa4775871ea8.png)
![new_laser_2](https://cloud.githubusercontent.com/assets/6357112/6318088/23ccc0e4-bae2-11e4-8d23-a5292e8a21e3.png)

Loganbacca commented:
Even though you only see the laser for a 3rd of a second, the end result of this change is actually quite dramatic since you can actually see where the laser is going.

This is also pretty cheap to run since your only doing 2 or 3 trig calculations, and the rest is just some basic math. There's also the advantage of the incremental movement costing nothing since its incrementing off pre-calculations.

Loganbacca commented on code/__HELPERS/vector.dm:
```diff
@@ -0,0 +1,151 @@
+              |
+              v
+             -90�
+             (W)
```
Should be ```(S)```

atlantiscze commented:
Is this purely visual (ie only smoothes what the game already did) or actually affects how projectiles are handled? (ie: Fixing the hyper-annoying bug where something, like turret, can shoot you, but you can't shoot it from the same position?)

Either way, very nice.

Zuhayr commented:
This is very cute.

mwerezak commented:
Definitely looks like an improvement.

Fox-McCloud commented:
Interesting will be keeping an eye on this.

That said, will probably require modification due to us not using instant beams.

Loganbacca commented:
@atlantiscze Yep, this fixes projectiles "curving". Instead they now travel in a true straight line to their target.

@Fox-McCloud The plot_vector datum should be able to be used with anything you want, just use our beams as an example and apply it to your own projectile code. It's designed to increment in single steps across the world, like step_towards() does.

PsiOmegaDelta commented:
Hopefully this will also fix turrets being unable to fire at objects at certain angles too.

Iamgoofball commented:
oh man I love it

laser50 commented:
Just gonna throw this in since it's been like this since a long time, any chance you can get it to spawn from one tile in front of the player instead of having the big gap? Dun seem legit

Loganbacca commented:
@laser50 I'm repurposing that one tile gap for muzzle flashes.

Fox-McCloud commented:
@Loganbacca 

If you're going to add in "muzzle flashes burn/blind you", you may want to make it a var so that downstream that doesn't want that (but wants the visual effect) can easily disable it.

Loganbacca commented:
@Fox-McCloud I didn't actually think of that, but I'll probably leave it as a purely visual effect anyway.

Fox-McCloud commented:
@Loganbacca 

Great! for the best, in my opinion.

RemieRichards commented:
Might this be useful as an alternative method for throwing items, as opposed to the current algorithm?

RemieRichards commented on code/modules/projectiles/projectile/beams.dm:
```diff
@@ -27,7 +31,10 @@ var/list/beam_master = list()
+
+			trajectory.increment()
+
+			step_towards(src, trajectory.return_location(get_turf(src))) //Move~
```
If I'm reading this right, isnt return_location() argumentless?
```/datum/plot_vector/proc/return_location()```

Loganbacca replied:
Yes, that's probably a holdover from an earlier version of ```plot_vector```.
Next thing I'll do with this is probably refactor plot_vector again as I've found I'll need more info from it for contiguous beams.

mwerezak commented:
Do you mind putting of some screenshots showing how the effect looks when the beam is blocked by an obstruction?

Loganbacca commented:
Currently the code is kinda in pieces while I rewrite ```plot_vector```, but basically it will terminate with an impact effect at the location it was blocked.

Loganbacca commented:
Updated. Most of the vector helper has been rewritten, and lasers and projectiles now both use ```plot_vector``` to move.
The effects associated with projectiles are currently in a state of being rewritten.

mwerezak commented:
Are you changing much besides the relative paths in projectile.dm?

Loganbacca commented:
Yes, I've changed process() a bit, but mostly just by adding the code for making projectiles travel in a true straight line.

One big change with this is I made bullets effectively hit-scan like lasers. I'll make a thread on the forums to confirm that's a change people are actually ok with.

Also, I'm almost done with this, I'm just finishing up on the effect code, but it's pretty much all working.

Fox-McCloud commented:
@Loganbacca 

How effectively will your code work with non-hitscan projectiles? I wouldn't be surprised if we (Para) and TG would both utilize this code, eventually, but if it can only be utilized with hitscan--well.

mwerezak commented:
@Loganbacca My main concern is conflicts with #8052.

Loganbacca commented:
@Fox-McCloud Pretty much just add a sleep back into process() to make it non-hitscan, I may even leave it in there commented out when I tidy this up.

@mwerezak This PR has been up for a month, you knew projectiles were being worked on. If yours is finished and pulled first, I'll have to adapt my code. If mine is finished first you'll have to adapt your code.

mwerezak commented:
I was hoping that if we're working on separate parts of projectile.dm then we could avoid most of the conflicts. It's hard to tell though with the changed paths and indentation.

Loganbacca commented:
@mwerezak #8073 should help once it's merged.

Loganbacca commented:
I'm still plodding away at this, but for the most part I'm going to wait for harpy to finish rewriting gun code before I finish it off.

mwerezak commented:
I've got some exams this week but hopefully(tm) the bulk of it will be done by the weekend.

Loganbacca commented:
(Force) pushed the latest updates.
Added a couple of screenshots to the original post showing the new effect (with codersprites).
Almost done.

Loganbacca commented on baystation12.int:
```diff
@@ -1,6 +1,11 @@
+WINDOW: icons\effects\projectiles.dmi
+FILE: icons\effects\projectiles.dmi
+*/
+// END_INTERNALS
```
Oops... will fix in next commit.

mwerezak commented on code/modules/projectiles/projectile.dm:
```diff
@@ -228,6 +240,10 @@
 	invisibility = 101
+
+	// generate impact effect
+	impact_effect(effect_transform)
```
Just be aware that many atom presently neglect to call on_hit().

It might make more sense to call this from on_impact()

RemieRichards commented on code/modules/projectiles/projectile.dm:
```diff
@@ -240,23 +256,95 @@
-					sleep(1)
-
+
+		if(hitscan)
```
You said NonHitscan weapons needed the sleep, so isn't this inverted?

Loganbacca replied:
Good point, I probably coded that way too late at night.

RemieRichards replied:
Been there, done that, programmed the t-shirt so it shows up with the "SUIT" icon...

ElArGee committed:
Projectile effects - Part 1

- Added base projectile effects framework
- Added vector movement helper
- Added temporary effect sprites

ElArGee committed:
Projectile effects - Part 2

- Converted the root projectiles object to support the new effects system at a generic level
- Converted laser beams to use the new effects system and removed the relics of the old system
- Added muzzle flash for ballistic ammo

ElArGee committed:
Projectile effects - Part 3

- Fixes for turrets, rogue drones, and emitters
- Removed muzzle flash from chem darts
- Converted emitters to use spanclasses and have correct capitalization

Loganbacca commented:
Ready for review/merge

comma commented:
:+1:

MistakeNot4892 commented:
I want to remind everyone subscribed to the Bay repo how fucking NICE this system is and was, ok thanks

RemieRichards commented:
Spook.

ElArGee commented:
❤️

Fox-McCloud commented:
3 Year old necro.

😱
